### PR TITLE
Add RDDCounts and visualization APIs, comprehensive documentation, and dual GNPS format support

### PR DESCRIPTION
**Pull request by AlejandroMC28**
Changed structure of the package(added classes for food counts and food flows), added visualization, utilities and testing

## Summary by CodeRabbit

* **New Features**
  * Added RDD counts computation, multi-level filtering, flow generation, PCA analysis (optional CLR), and multi-backend visualizations (Matplotlib & Plotly).

* **Documentation**
  * Added full docs, getting-started/tutorials, changelog, code of conduct, license, and README refresh.

* **Tests**
  * Comprehensive test suite covering init, filtering, flows, utils, visualization, and edge cases.

* **Chores**
  * CI and style workflows, packaging config added; removed an obsolete module and legacy setup script.

### CODE REVIEW

**Comment by bittremieux** [comment]
Can you remove the `__pycache__/` directories from the PR?

I recommend using a good global `.gitignore` file to avoid checking in such unnecessary files.

**Comment by AlejandroMC28** [comment]
Removed cache files and added a .gitignore file

File: gfop/analysis.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
+from sklearn.decomposition import PCA
```
Comment (bittremieux) [inline gfop/analysis.py]: **todo:** Structure import as follows.

- Three different groups, each separated by a blank line.
- The first group is for standard library imports.
- The second group is for third-party imports.
- The third group is for internal imports.
- In each group, `import ...` statements come before `from ... import ...`
- Within those subgroups, imports are sorted alphabetically.
- Prefer `import ...` over `from ... import ...` to have better self-documenting code (i.e. you'll easily see where each method comes from when reading the code).

File: gfop/analysis.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+from sklearn.preprocessing import StandardScaler
+from skbio.stats.composition import clr
+from foodcounts import FoodCounts
+from utils import food_counts_to_wide
+import pandas as pd
+import numpy as np
+
+
+def perform_pca_food_counts(
+    food_counts_instance,
```
Comment (bittremieux) [inline gfop/analysis.py]: **todo:** Add type hint.

File: gfop/analysis.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    level: int = 3,
+    n_components: int = 3,
+    apply_clr: bool = True,
+    food_types: list = None,
+):
```
Comment (bittremieux) [inline gfop/analysis.py]: **todo:** Add return type hint.

File: gfop/analysis.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    """
+    Perform PCA on food counts data using the FoodCounts instance, with an option for CLR transformation.
+
+    Args:
```
Comment (bittremieux) [inline gfop/analysis.py]: **todo:** We want to use NumPy style docstrings.

File: gfop/analysis.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+def perform_pca_food_counts(
+    food_counts_instance,
+    level: int = 3,
+    n_components: int = 3,
+    apply_clr: bool = True,
+    food_types: list = None,
+):
+    """
+    Perform PCA on food counts data using the FoodCounts instance, with an option for CLR transformation.
```
Comment (bittremieux) [inline gfop/analysis.py]: **todo:** Make sure to adhere to the maximum line width. Black will not automatically do this in (doc)strings.

File: gfop/analysis.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    Returns:
+        pd.DataFrame: A dataframe containing PCA scores, explained variance, and sample metadata.
+        list: Explained variance percentages.
+    """
+    # Step 1: Filter counts by level
+    food_counts_filtered = food_counts_instance.filter_counts(level=level)
+
+    # Step 2: Filter food types if specified by the user
+    if food_types:
```
Comment (bittremieux) [inline gfop/analysis.py]: **todo:** Test for `is not None`.

File: gfop/analysis.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        food_counts_filtered = food_counts_filtered[
+            food_counts_filtered["food_type"].isin(food_types)
+        ]
+
+    # Step 3: Convert to wide format for PCA
+    food_counts_wide = food_counts_to_wide(food_counts_filtered, level=level)
+
+    # Step 4: Extract numeric columns (food types) for PCA
+    food_type_columns = food_counts_wide.select_dtypes(include=[np.number]).columns
+    X = food_counts_wide[food_type_columns].values
```
Comment (bittremieux) [inline gfop/analysis.py]: **todo:** As per the Python convention, variables should start with a lowercase letter.

File: gfop/foodcounts.py [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+from importlib import resources
+import pkg_resources
+
+
+class FoodCounts:
+    def __init__(
+        self,
+        gnps_network: str,
+        sample_types: str,
+        all_groups: List[str],
```
Comment (bittremieux) [inline gfop/foodcounts.py]: **todo:** These two variables can probably be renamed to be more descriptive. It was crappy naming from me when I wrote the code.

File: gfop/foodcounts.py [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+        self._validate_groups()
+
+        # Generate sample metadata and counts
+        self.sample_metadata = self._get_sample_metadata()
+        self.counts = self.create()
+
+    def _validate_groups(self) -> None:
+        """
+        Validates that the provided group names exist in the GNPS network data.
+        Raises a ValueError if invalid group names are found.
```
Comment (bittremieux) [inline gfop/foodcounts.py]: **todo:** Use the proper docstring formatting to describe the error.

File: gfop/foodcounts.py [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+        """
+        valid_groups = set(self.gnps_network["DefaultGroups"].unique())
+
+        # Check all_groups
+        invalid_all_groups = set(self.all_groups) - valid_groups
+        if invalid_all_groups:
```
Comment (bittremieux) [inline gfop/foodcounts.py]: **todo:** `if len(invalid_all_groups) > 0`.

File: gfop/foodcounts.py [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+            raise ValueError(
+                f"The following groups in all_groups are invalid: {invalid_all_groups}"
+            )
+
+        # Check some_groups
+        invalid_some_groups = set(self.some_groups) - valid_groups
+        if invalid_some_groups:
```
Comment (bittremieux) [inline gfop/foodcounts.py]: **todo:** Same here.

File: gfop/foodcounts.py [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+            ) as stream:
+                gfop_metadata = pd.read_csv(stream, sep="\t")
+        except (ModuleNotFoundError, ImportError):
+            stream = pkg_resources.resource_stream(
+                __name__, "data/foodomics_multiproject_metadata.txt"
+            )
+            gfop_metadata = pd.read_csv(stream, sep="\t")
+
+        # Remove trailing whitespace
+        gfop_metadata = gfop_metadata.apply(
```
Comment (bittremieux) [inline gfop/foodcounts.py]: **todo:** The metadata is packaged with the code, so just modify the file that it is fully correct.

File: gfop/foodcounts.py [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+        """
+        food_counts, filenames = [], []
+        metadata = self.sample_metadata
+
+        for filename in metadata["filename"]:
+            file_food_counts = self._get_file_food_counts(level, [filename])
+            if len(file_food_counts) > 0:
+                food_counts.append(file_food_counts)
+                filenames.append(filename)
+        if not food_counts:
```
Comment (bittremieux) [inline gfop/foodcounts.py]: **todo:** Test using `len`.

File: gfop/foodflows.py [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+        self.sample_types = self._load_sample_types(sample_types)
+        self.groups_included = groups_included
+        self.max_hierarchy_level = max_hierarchy_level
+        self.food_metadata = self._load_food_metadata()
+        # Validate group names
+        self._validate_groups()
+        # Generate flows and processes dataframes
+        self.flows, self.processes = self.generate_foodflows()
+
+    def _validate_groups(self) -> None:
```
Comment (bittremieux) [inline gfop/foodflows.py]: **question:** The next few methods seem identical as in `foodcounts.py`. Can this code duplication be avoided by reusing functionality from there?

File: gfop/notebooks/01_food_counts_notebook.ipynb
Comment (bittremieux) [inline gfop/notebooks/01_food_counts_notebook.ipynb:1]: **todo:** Remove the notebooks from the code for now. We'll discuss later how to include documentation and code examples.

File: gfop/utils.py [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+        pd.DataFrame: A wide-format dataframe where each combination of 'food_type' and 'group' is a column.
+    """
+    # If level is not specified, infer the level from the data
+    if level is None:
+        levels_in_data = food_counts["level"].unique()
+        if len(levels_in_data) > 1:
+            raise ValueError(
+                "Multiple levels found in the data. Please specify a level to convert to wide format."
+            )
+        level = levels_in_data[0]  # If the data is already filtered, use that level
```
Comment (bittremieux) [inline gfop/utils.py]: **praise:** Nice.

File: gfop/visualization.py [hunk 16]
```diff
@@ -0,0 +1,582 @@
...
+import plotly.express as px
+from typing import List
+from foodcounts import FoodCounts
+from utils import food_counts_to_wide, calculate_proportions
+import pandas as pd
+import plotly.graph_objects as go
+
+
+def plot_food_type_distribution(
+    food_counts_instance,
```
Comment (bittremieux) [inline gfop/visualization.py]: **todo:** Type hint.

**Review by bittremieux — CHANGES_REQUESTED** [review]
Nice work. I left a few comments to mainly update some of the code style a bit. Note that some of these comments re-occur in a few different places but I haven't marked all of them, so go through the code with them in mind.

File: gfop/analysis.py (hunk 1 above)
Comment (AlejandroMC28) [reply gfop/analysis.py]: Resolved for all modules

**Comment by AlejandroMC28** [comment]
Resolved all comments, updated the base algorithm of food counts computation.

File: gfop/analysis.py [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+
+# Standard library imports
+from typing import List, Tuple, Optional
+
+# Third-party imports
+import numpy as np
+import pandas as pd
+from sklearn.decomposition import PCA
+from sklearn.preprocessing import StandardScaler
+from skbio.stats.composition import clr
```
Comment (bittremieux) [inline gfop/analysis.py:11]: **nitpick:** Alphabetical order.

File: gfop/foodcounts.py [hunk 18]
```diff
@@ -0,0 +1,339 @@
...
+        filenames_df = df_exploded_files[["DefaultGroups", "UniqueFileSources"]].rename(
+            columns={"DefaultGroups": "group", "UniqueFileSources": "filename"}
+        )
+        return filenames_df.drop_duplicates().reset_index(drop=True)
+
+    def _get_filename_level_food_counts(self) -> pd.DataFrame:
+        """
+        Generates a table of food counts at the filename level. It filters the GNPS network
+         based on the provided sample and
+        reference groups.
```
Comment (bittremieux) [inline gfop/foodcounts.py]: **nitpick:** Newlines fix.

File: gfop/foodflows.py [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+        Generates food flows and processes from the GNPS network data.
+
+        Returns
+        -------
+        Tuple[pd.DataFrame, pd.DataFrame]
+            A tuple containing:
+            - flows: DataFrame with source, target, and value columns.
+            - processes: DataFrame with process details.
+        """
+        # Select GNPS job groups.
```
Comment (bittremieux) [inline gfop/foodflows.py]: **question:** This first part repeats some behavior of the `FoodCounts` class. Can `FoodFlows` instead take a `FoodCounts` object as input on which these steps are already applied, and then only take care of the visualization?

File: gfop/foodflows.py [hunk 20]
```diff
@@ -0,0 +1,122 @@
+# Standard library imports
+import os
+import pkg_resources
+from importlib import resources
```
Comment (bittremieux) [inline gfop/foodflows.py]: **question:** Not used?

File: gfop/visualization.py [hunk 21]
```diff
@@ -0,0 +1,674 @@
+# visualization.py
+
+# Standard library imports
+import os
+from typing import List, Optional, Union, Tuple
+from importlib import resources
```
Comment (bittremieux) [inline gfop/visualization.py]: **nitpick:** Alphabetical order.

File: gfop/visualization.py [hunk 22]
```diff
@@ -0,0 +1,674 @@
...
+
+# Third-party imports
+import matplotlib.pyplot as plt
+import matplotlib.figure
```
Comment (bittremieux) [inline gfop/visualization.py]: **nitpick:** Alphabetical order.

File: gfop/visualization.py [hunk 23]
```diff
@@ -0,0 +1,674 @@
...
+import seaborn as sns
```
Comment (bittremieux) [inline gfop/visualization.py]: **nitpick:** Alphabetical order.

File: tests/test_foodcounts.py [hunk 24]
```diff
@@ -0,0 +1,254 @@
+# Standard library imports
+import os
+import sys
+
+# Third-party imports
+import pytest
```
Comment (bittremieux) [inline tests/test_foodcounts.py]: **nitpick:** Alphabetical order.

File: tests/test_foodcounts.py [hunk 25]
```diff
@@ -0,0 +1,254 @@
...
+import pandas as pd
+
+# Add the path to the 'gfop' directory to the system path
+project_path = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "gfop"))
+sys.path.append(project_path)
+
+# Internal imports
+from foodcounts import FoodCounts  # type: ignore
+
+
```
Comment (bittremieux) [inline tests/test_foodcounts.py]: **nitpick:** Too many blank lines.

File: tests/test_foodcounts.py
Comment (bittremieux) [inline tests/test_foodcounts.py:1]: **todo:** My key concern is that several large files are added now for the tests. This will lead to considerable bloat for the package. Instead, a simple molecular networking file can be simulated in a fixture to use for testing. The actual values don't matter that much, it doesn't need to be real data, it just needs to work to test whether the code operates correctly.

**Review by bittremieux — CHANGES_REQUESTED** [review]
The code is starting to look very nice, good job. I have one main concern, and that's the addition of several big(ish) sample files used for testing. This should ideally be avoided, see the comment in the code.

**Comment by AlejandroMC28** [comment]
Solved previous comments, and 8/11 issues. Issues remaining are to add changelog, documentation and contributing guidelines

**Review by bittremieux — APPROVED** [review]
